File: 231119/악수와 전염병의 상관관계 2/correlation-between-shaking-hands-and-infectious-diseases2.py
N,K,P,T=map(int,input().split())

MAX_T=250*250
people=[[0 for j in range(2)] for i in range(N+1)]
time=[(0,0)]*(MAX_T+1)
people[P][0]=1

maxt=0
for _ in range(T):
    t,x,y=map(int,input().split())
    time[t]=(x,y)
    maxt=max(t,maxt)

for t in range(1,maxt+1):
    x,y=time[t]

    if people[x][0]==1 and people[x][1]<K:
        people[y][0]=1
        people[x][1]+=1
    elif people[y][0]==1 and people[y][1]<K:
        people[x][0]=1
        people[y][1]+=1


for t in range(1,N+1):
    print(people[t][0],end='')

File: 251209/악수와 전염병의 상관관계 2/correlation-between-shaking-hands-and-infectious-diseases2.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const [N,K,P,T] = input[0].split(' ').map(Number);
const shakes = input.slice(1).map((shake)=>shake.split(' ').map(Number));

shakes.sort((a,b)=>a[0]-b[0]);
const maxT = shakes[T-1][0];

const arr = Array(N+1).fill(0);
const cnt = Array(N+1).fill(0);

arr[P]=1;
cnt[P]=K;

for(let shake of shakes){
    let [t, a, b] = shake;
    if(cnt[a]>0 && cnt[b]>0){
        cnt[a]--;
        cnt[b]--;
    }else if(cnt[a]>0){
        cnt[a]--;
        if(arr[b]!==1){
            arr[b]=1;
            cnt[b]=K;
        }
    }else if(cnt[b]>0){
        cnt[b]--;
        if(arr[a]!==1){
            cnt[a]=K;
            arr[a]=1;
        }
    }
}

arr.shift()
console.log(arr.join(''))